examples_tests/28.BSDFValidator/BSDFValidatorApp.cpp

```diff
@@ -50,34 +53,125 @@ BSDFValidatorApp::BSDFValidatorApp(IrrlichtDevice& device)
         root->getChild("LoadDefinitionsButton"));
     loadDefinitionsButton->subscribeEvent(::CEGUI::PushButton::EventClicked,
         ::CEGUI::Event::Subscriber(&BSDFValidatorApp::EventFunctionDefinitionBrowse, this));
+
+    /* !!!....TEMPORARY....!!! */
+    // This whole mesh business is just to check if the custom shaders
+    // are woring as intended. It will not make it to the final version.
+#include "irr/irrpack.h"
+    struct VertexStruct
+    {
+        float position[3]; /// uses float hence need 4 byte alignment
+    } PACK_STRUCT;
+#include "irr/irrunpack.h"
+
+    // Create mesh.
+    VertexStruct vertices[4] =
+    {
+        { -0.5f, -0.5f, 0.f },
+        { 0.5f, -0.5f, 0.f },
+        { 0.5f, 0.5f, 0.f },
+        { -0.5f, 0.5f, 0.f }
+    };
+
+    uint16_t indices[6] =
+    {
+        0, 1, 2,
+        2, 3, 0
+    };
+
+    // Upload mesh to the GPU
+    auto upStreamBuff = m_VideoDriver->getDefaultUpStreamingBuffer();
+    const void* dataToPlace[2] = { vertices, indices };
+    uint32_t offsets[2] = { video::StreamingTransientDataBufferMT<>::invalid_address,video::StreamingTransientDataBufferMT<>::invalid_address };
+    uint32_t alignments[2] = { sizeof(decltype(vertices[0u])),sizeof(decltype(indices[0u])) };
+    uint32_t sizes[2] = { sizeof(vertices),sizeof(indices) };
+    upStreamBuff->multi_place(2u, (const void* const*)dataToPlace, offsets, sizes, alignments);
+
+    if (upStreamBuff->needsManualFlushOrInvalidate())
+    {
+        auto upStreamMem = upStreamBuff->getBuffer()->getBoundMemory();
+        m_VideoDriver->flushMappedMemoryRanges({ video::IDriverMemoryAllocation::MappedMemoryRange(upStreamMem, offsets[0],sizes[0]),video::IDriverMemoryAllocation::MappedMemoryRange(upStreamMem,offsets[1],sizes[1]) });
+    }
+
+    m_Mesh = core::make_smart_refctd_ptr<video::IGPUMeshBuffer>();
+
+    {
+        // Define vertex attribute layout. 
+        auto desc = m_VideoDriver->createGPUMeshDataFormatDesc();
+
+        {
+            auto buff = core::smart_refctd_ptr<video::IGPUBuffer>(upStreamBuff->getBuffer());
+
+            // Position attribute.
+            desc->setVertexAttrBuffer(core::smart_refctd_ptr(buff), asset::EVAI_ATTR0, asset::EF_R32G32B32_SFLOAT, sizeof(VertexStruct), 0);
+
+            desc->setIndexBuffer(std::move(buff));
+        }
+
+        m_Mesh->setIndexBufferOffset(offsets[1]);
+        m_Mesh->setIndexType(asset::EIT_16BIT);
+        m_Mesh->setIndexCount(6);
+        m_Mesh->setMeshDataAndFormat(std::move(desc));
+    }
+
+    m_VideoDriver->setTransform(video::E4X3TS_WORLD, core::matrix4x3());
+
+    upStreamBuff->multi_free(2u, (uint32_t*)&offsets, (uint32_t*)&sizes, m_VideoDriver->placeFence());
+
+}
+
+BSDFValidatorApp::~BSDFValidatorApp()
+{
+    delete m_ShaderManager;
```

Comment:
you could use a `smart_refctd_ptr` you know, if only `irr::ShaderManager` inherited from `irr::core::ReferenceCounted`